python/cudf/dataframe/dataframe.py

```diff
@@ -469,18 +469,25 @@ def take(self, positions, ignore_index=False):
             out[col] = self[col].take(positions, ignore_index=ignore_index)
         return out
 
-    def copy(self):
+    def copy(self, deep=True):
         """
         Returns a copy of this dataframe
+
+        Parameters
+        ----------
+        deep: bool
+           Make a full copy of Series columns and Index at the GPU level, or
+           create a new allocation with references.
         """
         df = DataFrame()
-        df._index = self._index
         df._size = self._size
-        df._cols = self._cols.copy()
+        for k in self._cols:
+            df._index = self._index.copy(deep)
+            df._cols[k] = self._cols[k].copy(deep)
```

Comment:
If a shallow copy it should return the same instances of the index and series objects, not shallow copies of them.